### PR TITLE
Fix: track variables, not names in require-atomic-updates (fixes #14208)

### LINKED ISSUE #14208: [require-atomic-updates]: False positive on reused loop variable name
**Tell us about your environment**

Environment Info:

Node version: v14.15.4
npm version: v7.5.3
Local ESLint version: v7.21.0 (Currently used)
Global ESLint version: Not found

Parser: `@typescript-eslint/parser`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  env: {
    es2020: true,
    node: true
  },
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: {
    'project': './tsconfig.eslint.json',
    'sourceType': 'module',
    'tsconfigRootDir': __dirname,
  },
  plugins: [
    'eslint-plugin-import',
    '@typescript-eslint',
  ],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:@typescript-eslint/recommended-requiring-type-checking',
  ],
  rules: {
    '@typescript-eslint/no-non-null-assertion': 'warn',
    '@typescript-eslint/no-shadow': ['error', { 'hoist': 'all' }],
    '@typescript-eslint/no-unused-expressions': 'error',
    '@typescript-eslint/require-await': 'warn',
    '@typescript-eslint/require-array-sort-compare': 'error',
    '@typescript-eslint/no-base-to-string': 'error',
    '@typescript-eslint/restrict-plus-operands': 'error',
    'array-callback-return': 'error',
    'eqeqeq': ['error', 'always'],
    'guard-for-in': 'error',
    'no-bitwise': 'error',
    'no-caller': 'error',
    'no-eval': 'error',
    'no-new-wrappers': 'error',
    'no-template-curly-in-string': 'error',
    'no-throw-literal': 'error',
    'no-undef-init': 'error',
    'no-unreachable-loop': 'error',
    'no-useless-backreference': 'error',
    'no-var': 'error',
    'require-atomic-updates': 'error',
    'prefer-const': 'off',      // change to 'error'
    '@typescript-eslint/consistent-type-definitions': ['warn', 'interface'],
    '@typescript-eslint/semi': ['error', 'always'],
    '@typescript-eslint/unified-signatures': 'error',
    '@typescript-eslint/prefer-string-starts-ends-with': 'warn',
    '@typescript-eslint/no-extra-semi': 'off',        // change to 'warn'
    '@typescript-eslint/no-inferrable-types': 'off',    // change to ['warn', {ignoreParameters: true}]
    'arrow-body-style': 'off',          // change to ['error', 'as-needed'],
    '@typescript-eslint/no-explicit-any': 'off',
    '@typescript-eslint/explicit-module-boundary-types': 'off',
    '@typescript-eslint/no-unused-vars': 'off',
    '@typescript-eslint/no-empty-function': 'off',
    '@typescript-eslint/no-unsafe-assignment': 'off',
    '@typescript-eslint/no-unsafe-member-access': 'off',
    '@typescript-eslint/no-unsafe-call': 'off',
    '@typescript-eslint/no-unsafe-return': 'off',
    '@typescript-eslint/no-implicit-any-catch': 'off',
    '@typescript-eslint/restrict-template-expressions': 'off',
    '@typescript-eslint/no-unnecessary-condition': 'off',
    'no-inner-declarations': 'off',
    'no-promise-executor-return': 'off',
  }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
async function foo(e) {
}

async function run() {
  const input = [];
  const props = [];

  for(const entry of input) {
    const prop = props.find(a => a.id === entry.id) || null;
    await foo(entry);
  }

  // Possible race condition: `const entry` might be reassigned based on an outdated value of `const entry`
  for(const entry of input) {
    const prop = props.find(a => a.id === entry.id) || null;
  }

  for(const entry2 of input) {
    const prop = props.find(a => a.id === entry2.id) || null;
  }
}
```

Command: `npx eslint scripts/sample-require-atomic-updates.ts`

**What did you expect to happen?**

Did not expect any errors because the `entry` variables in the two loops are independent.

**What actually happened? Please include the actual, raw output from ESLint.**

ESLint seems to think that the variable captured in the second loop's arrow function closure is the `entry` variable from the first loop, and flags an error.

```
/......./scripts/sample-require-atomic-updates.ts
  14:3  error  Possible race condition: `const entry` might be reassigned based on an outdated value of `const entry`  require-atomic-updates

✖ 1 problem (1 error, 0 warnings)
```

### PR DESCRIPTION
**Pull request by patriscus**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

fixes #14208

Fixes false positives like:
```js
async function foo(e) {
}

async function run() {
  const input = [];
  const props = [];

  for(const entry of input) {
    const prop = props.find(a => a.id === entry.id) || null;
    await foo(entry);
  }

  // false positive:
  // Possible race condition: `const entry` might be reassigned based on an outdated value of `const entry`
  for(const entry of input) {
    const prop = props.find(a => a.id === entry.id) || null;
  }

  for(const entry2 of input) {
    const prop = props.find(a => a.id === entry2.id) || null;
  }
}
```

#### What changes did you make? (Give an overview)

I adapted the `if` condition surrounding the code block that is pushing/setting a reference to the `assignmentReferences` map. This map is subsequently used to get `references` and check for outdated identifier names.

I moved the `isLocalVariableWithoutEscape` in said `if` condition out to a separate variable and extended it with: 
```js
const isVariableDeclaration = isLocalVariableWithoutEscape(variable, isMemberAccess) ||
    node.parent.type === "VariableDeclarator";
```

Since the condition had a comment with  "exclude variable declarations", I removed it and expressed it with the `isVariableDeclaration` variable name.

#### Is there anything you'd like reviewers to focus on?

This is the first time contributing to ESLint (or open-source that is). Please focus on edge-cases or other implications since I do not have the complete context. I'm happy to hear about your concerns and improvement points.

### CODE REVIEW

**Review by mdjermanovic — CHANGES_REQUESTED** [review]
@patriscus thanks for the PR! I think that this solution does fix #14208, but not all similar cases, and maybe it wouldn't be too difficult to fix the actual root cause.

File: lib/rules/require-atomic-updates.js [hunk 1]
```diff
@@ -224,13 +224,16 @@ module.exports = {
...
+
                 /*
                  * Register the variable to verify after ESLint traversed the `writeExpr` node
                  * if this reference is an assignment to a variable which is referred from other closure.
                  */
                 if (writeExpr &&
-                    writeExpr.parent.right === writeExpr && // ← exclude variable declarations.
-                    !isLocalVariableWithoutEscape(variable, isMemberAccess)
+                    writeExpr.parent.right === writeExpr &&
+                    !isVariableDeclaration
```
[the remark this answers is not in the document]
Comment (patriscus) [reply lib/rules/require-atomic-updates.js:236]: Thanks for the review and the tip! I'll try to get to it in the upcoming week and will give an update latest next weekend

**Comment by patriscus** [comment]
@mdjermanovic I had some spare time and changed the code such that `segmentInfo` stores references to variables instead of storing variable names. Feel free to review & leave suggestions again. Thanks!

**Review by mdjermanovic — CHANGES_REQUESTED** [review]
The change looks good, we should just skip unresolved references in `function createReferenceMap`.

`reference.resolved` is `null` if it's unresolved, so this would be a false positive:

```js
/* eslint require-atomic-updates: error */

async function run() {
    await a;
    b = 1; // false positive
}
```

(because we're now tracking undefined variables `a` and `b` as the same `null` variable object in `SegmentInfo`).

That would miss reporting some cases with the same variable name, but it's fine, most of the core rules are ignoring undefined variables.

**Comment by patriscus** [comment]
I'll get to this PR hopefully this weekend, latest next weekend

**Comment by patriscus** [comment]
@mdjermanovic I have now updated the code to skip unresolved references. Please let me know if you have further suggestions.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Comment by patriscus** [comment]
Thanks for guiding me through this PR! Much appreciated

**Review by btmills — APPROVED** [review]
LGTM. Congrats on your first open source PR, @patriscus! This will go out in tonight's release.

**Pull request merged**